doc/memos/rdm-draft-alamos-lanzieri-runtime-configuration-architecture.md

```diff
@@ -0,0 +1,139 @@
+- Export configuration parameters (e.g copy to a buffer, print, etc)
+
+The RIOT registry consists of:
+- A list of Registry Handlers
+- A list of one or more Storage Facilities sources
+- One Storage destination
+
+# 4. Registry handlers
+A Registry Handler represents a configuration group in the RIOT Registry.
+Examples of configuration groups could be _Application_, _LoRaWAN keys_, etc.
+
+A Registry Handler is defined at least by a name and a series of handlers for interacting with the configuration parameters of the group. These handlers are:
+- `set`: Sets a value to a configuration parameter. This handler MUST take care
+of any logic to apply the value (e.g. data validation) or cache it until the commit handler is called
+- `get`: Gets the current value of a configuration parameter
+- `commit`: To be called when configuration parameters have been loaded from storage. It could be use for special needed logic on applying configuration parameters (e.g dependencies)
+- `export`: Calls an `export function` for each configuration parameter, with
+its name and value. Depending on the behavior of `export function`, this can be used for printing out all configurations, saving them in a persistent storage,
+etc.
+
+# 5. Storage facilities
+Storage facilities MUST implement the **storage
+interface** to allow the RIOT Registry to load, search and save configuration
+parameters. From the point of view of the RIOT Registry all parameters are
+key/value strings, it is responsibility of the storage facility to transform
+that to the proper format for storage (e.g. lines separated by `\n` character in
+a file).
+
+If a storage implements any kind of encryption mechanism it MUST be
+transparent to the RIOT Registry, and all that logic should be handled withing
+the facility.
+
+# 6. Configuration manager and interfaces
+The configuration manager is a pseudo-module that allows a RIOT node to be
+configured from one or more communication interfaces. Examples of these
+communication interfaces could be UART, SPI or higher layers like PPP, IPv6,
+UDP, CoAP, LWM2M, etc.
+
+The configuration manager MAY provide an access control mechanism for
+restricting access to the configurations. However, it can relay on security
+implemented on the layers mentioned above (e.g CoAP and DTLS, CHAP on PPP, etc)
+
+# 7. RCS Usage Flow
+1. Modules declare and register **registry handlers** for configuration groups
+in the RIOT Registry.
+
+2. **Storage facilities are registered** as sources and/or destinations of configurations in the RIOT Registry.
+
+3. The RIOT Registry is initiated. All configuration parameters are loaded from the sources with the Storage interface.
+
+4. For each loaded parameter, a configuration group that contains that parameter is searched for. If found, the corresponding `set` handler is called with the current parameter and value.
+
+5. The `commit` handler is called once all configurations have been loaded from the registered sources.
```

Comment:
Does this relate to some form of transactional processing?

Reply:
yes, I think that's the proper concept.